### PR TITLE
support case-insensitive query options in SET syntax

### PR DESCRIPTION
**Pull request by agavra**
allows users to specify case-insensitive options when using `SET` syntax to be more similar to what SQL users expect. Note that this does not apply to the `queryOptions` JSON in the request, that will need to be a follow-up PR.

Tested e2e using Quickstart (`SET useMULTIstageEngine=true; select * from billing limit 10;`)

### CODE REVIEW

**Review by walterddr — COMMENTED** [review]
could you rebase, looks like the integration test had some issues so need to retrigger a GHA for a fresher base biuild

**Comment by agavra** [comment]
rebased

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/QueryOptionsUtils.java [hunk 1]
```diff
@@ -32,6 +36,59 @@ public class QueryOptionsUtils {
...
+
+    Map<String, String> resolved = new HashMap<>();
+    for (Map.Entry<String, String> configEntry : queryOptions.entrySet()) {
+      String config = CONFIG_RESOLVER.get(configEntry.getKey().toLowerCase());
+      if (config != null) {
+        resolved.put(config, configEntry.getValue());
+      } else {
+        resolved.put(configEntry.getKey(), configEntry.getValue());
+      }
+    }
```
Comment (walterddr) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/QueryOptionsUtils.java:87]: seems like the reason why this has to be so complex is b/c `QueryOptionKey` is a inner class not a enum
it looks like thing will get much simpler if we can do the same as `DataTable.MetadataKey` which is an ENUM. 

@siddharthteotia any concerns on backward compatibility?

File: pinot-spi/src/test/java/org/apache/pinot/spi/utils/QueryOptionsUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+public class QueryOptionsUtilsTest {
+
+  @Test
+  public void shouldConvertCaseInsensitiveMapToUseCorrectValues() {
+    // Given:
+    Map<String, String> configs = ImmutableMap.of(
+        "ENABLENullHandling", "true",
+        "useMULTISTAGEEngine", "false"
+    );
```
Comment (walterddr) [inline pinot-spi/src/test/java/org/apache/pinot/spi/utils/QueryOptionsUtilsTest.java:36]: let's also add a test for option key-value pairs that are not in the QueryOptionKey set
1. with underscores
2. with partial match
3. completely irrelevant. 
4. dot separated and some escaped characters. 

the goal is to show that only those in QueryOptionKey are case insensitive. others are not

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java [hunk 3]
```diff
@@ -171,7 +172,7 @@ static SqlNodeAndOptions extractSqlNodeAndOptions(String sql, SqlNodeList sqlNod
     if (sqlType == null) {
       throw new SqlCompilationException("SqlNode with executable statement not found!");
     }
-    return new SqlNodeAndOptions(statementNode, sqlType, options);
+    return new SqlNodeAndOptions(statementNode, sqlType, QueryOptionsUtils.resolveCaseInsensitiveOptions(options));
```
Comment (walterddr) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:175]: this has several issues
1. using legacy OPTION keyword will still be case-sensitive; and legacy OPTIONS keyword is written into the map after resolving SqlNodeAndOptions; how do we resolve conflicts in this situation? for example:
```
SET useMultistageengine = true;
SELECT ...
OPTION(useMultistageEngine, false)
```
2. SqlNodeAndOption only parses options in query, how does it handle json payload option keys?

**Comment by walterddr** [comment]
> @walterddr I'm not sure I agree with the concerns - for the first this will still work just fine; the second makes sense because SQL supports case-sensitivity (generally they would require you to quote case sensitive names, but at least it's supported). I've never seen a configuration system with case sensitive configs. It's kind of like requiring the SQL keywords to be case sensitive (e.g. `SELECT *` would work but `select *` wouldn't)

sound good. as long as we limited the list of case insensitive options to the known ones in `QueryOptionKey` --> which seems logical compare to the `SELECT *` vs. `select *` analogy (reserved keywords are always case insensitive)

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java (hunk 3 above)
Comment (agavra) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:175]: I considered both of these and thought it was OK to make this a partial improvement. the SET syntax is particularly misleading given that it "feels" like sql - if we feel strongly about it I can attempt to wire this in through the other places as well.

Comment (walterddr) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:175]: it should be fine. but let's make sure to test when the other 2 ways to configure options are present - and what the desired behavior is - and that should match the behavior of current logic.

Comment (walterddr) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:175]: e.g. also test
1. JSON option `timeoutMs = 60000`
2. query option `OPTION(timeoutMs=30000)`
3. query set `SET timeoutMs = 20000`
4. implicitly there's a default 10000

let's make sure the combination of these result in the same final timeout number
- by setting 1,2,3 to `timeoutms` or `TIMEOUTMS` or `timeoutMs`

Comment (agavra) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:175]: looks like there was actually a pretty easy way to make all of them work :) I'll update the PR

Comment (agavra) [reply pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:175]: confirmed with the following:
```
select * from dates limit 10; #no multistage default
select * from dates limit 10 OPTION(usemultistageengine=true) #use options
SET usemultistageengine=true; select * from dates limit 10;
```
and also via curl for json:
```
curl 'http://localhost:9000/sql' --data-raw '{"sql":"select * from dates limit 10","trace":false,"queryOptions":"usemultistageengine=true"}'
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/QueryOptionsUtils.java [hunk 4]
```diff
@@ -16,9 +16,13 @@
  * specific language governing permissions and limitations
  * under the License.
  */
-package org.apache.pinot.core.util;
+package org.apache.pinot.spi.utils;
```
Comment (walterddr) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/QueryOptionsUtils.java:19]: do we need to move this Util to SPI? looks like the only place it is being used is common. cant we move it into common?

**Review by walterddr — COMMENTED** [review]
lgtm overall. one more comment

**Review by walterddr — APPROVED** [review]
lgtm. please fix linter, thanks

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/QueryOptionsUtils.java (hunk 1 above)
Comment (siddharthteotia) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/QueryOptionsUtils.java]: @walterddr @agavra  - I will review this today

**Review by siddharthteotia — APPROVED** [review]
Looks good to me. I don't think this changes the way options are passed so we should be good. Just collapses the options ignoring case.

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/QueryOptionsUtils.java (hunk 1 above)
Comment (siddharthteotia) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/QueryOptionsUtils.java]: Looks good to me.

**Pull request merged**